Add implicit usings property, and global usings page

adamint opened the pull request:
This PR adds the new global usings property page. It moves the checkbox to enable global usings, and conditions the usings editor on that value being true.

You are allowed to add an implicit using - the "is static" checkbox is disabled if an alias of any length is added.

You are allowed to edit non-readonly implicit usings (currently, usings imported from props/sdk), or delete them. 

Adding validation to this feature is tracked in #8439

![image](https://user-images.githubusercontent.com/20359921/187303882-457f91c9-f5b0-44fc-9030-5f9b7026f983.png)


note: this exposed a bug I introduced in CPS where focus is removed after collectionview refresh. if testing, please clone CPS off this PR: https://devdiv.visualstudio.com/DevDiv/_git/CPS/pullrequest/420346

Closes #7488 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/8438)

melytc commented:
So, I've started reviewing it, but oof, is quite a long PR with lots of files changed. I know some feature work needs to be done in one take rather than in various ones; it might be helpful for next time to have a more descriptive commit history, where each commit does one thing (or aims to 😉), and you document your changes.

adamint commented:
The highlights:
- Copied PropertyEditorBase/related classes from CPS into project-system to allow us to add editor resources like in PropertyTemplates.xaml
- Added a bunch of converters from CPS
- Added KeyboardFocusBehaviour from CPS

Everything else is new code.
- There's a new usings value provider, as well as the CSharpImplicitUsingsControl,which actually make up the new control.

melytc commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ReferencesPage/ImplicitUsingsValueProvider.cs:
```diff
@@ -0,0 +1,200 @@
+        return null;
+    }
+
+    internal class ImplicitUsing
```
Could this class be promoted to be its own file? It might help specially when adding the VB experience, I suspect we would re-use this.

tmeschter replied:
Bonus points for turning this into a C# `record`. ;-)

adamint replied:
We target <net5 in this project, so we run into the following error: The predefined type 'System.Runtime.CompilerServices.IsExternalInit' must be defined or imported in order to declare init-only setter

Unfortunately, we are not able to use record types without a hacky workaround.

melytc commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/DotNetImportsEnumProvider.cs:
```diff
@@ -7,7 +7,7 @@
 
 [ExportDynamicEnumValuesProvider("DotNetImportsEnumProvider")]
-[AppliesTo(ProjectCapability.VisualBasic)]
+[AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
```
I can see this `DotNetImportsEnumProvider` is used in the ReferencesPage.VisualBasic.xaml; where is it being referenced for the C# references?

melytc reviewed (commented):
Some initial thoughts/suggestions

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/PropertyPageResources.resx:
```diff
@@ -279,4 +279,28 @@
   </data>
+  <data name="ImplicitUsingEditor_AddUsing" xml:space="preserve">
+    <value>Using</value>
+  </data>
```
It would be very useful to have comments for each of these explaining what the text means and/or where it is used. Our translators generally see these terms in isolation and the more context we can provide, the better the translation.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/PropertyPageResources.resx:
```diff
@@ -279,4 +279,28 @@
+    <value>Alias</value>
+  </data>
+  <data name="ImplicitUsingEditor_ListItemIsStatic" xml:space="preserve">
+    <value>Is Static</value>
```
I would prefer just "Static" over "Is Static".

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/xlf/PropertyPageStrings.cs.xlf:
```diff
@@ -0,0 +1,12 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<xliff xmlns="urn:oasis:names:tc:xliff:document:1.2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="1.2" xsi:schemaLocation="urn:oasis:names:tc:xliff:document:1.2 xliff-core-1.2-transitional.xsd">
+  <file datatype="xml" source-language="en" target-language="cs" original="../PropertyPageStrings.resx">
```
Were these PropertyPageStrings.*.xlf files accidentally committed? I don't see a PropertyPageStrings.resx in the change.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Microsoft.VisualStudio.ProjectSystem.Managed.VS.csproj:
```diff
@@ -52,15 +52,6 @@
-      <AutoGen>True</AutoGen>
-      <DesignTime>True</DesignTime>
-      <DependentUpon>PropertyPageResources.resx</DependentUpon>
-    </Compile>
```
Were these `Update` items just not needed?

MiYanni replied:
Even though these files are checked into the repo, they are also generated. These updates should likely not be removed.

adamint replied:
No, it didn't look like it was needed, but I will re-add

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/PublicAPI.Shipped.txt:
```diff
@@ -416,4 +416,12 @@ virtual Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies.AttachedCollec
+~static Microsoft.VisualStudio.ProjectSystem.VS.PropertyPages.PropertyPageResources.ImplicitUsingEditor_ListItemIncludes.get -> string
+~static Microsoft.VisualStudio.ProjectSystem.VS.PropertyPages.PropertyPageResources.ImplicitUsingEditor_ListItemIsStatic.get -> string
+~static Microsoft.VisualStudio.ProjectSystem.VS.PropertyPages.PropertyPageResources.ImplicitUsingEditor_ListViewAccessibleName.get -> string
+~static Microsoft.VisualStudio.ProjectSystem.VS.PropertyPages.PropertyPageResources.ImplicitUsingEditor_ShowImportedUsingsButton.get -> string
```
I take it these are public because WPF needs them to be public?

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/Converters/Converters.cs:
```diff
@@ -0,0 +1,101 @@
+
+        public object[] ConvertBack(object value, Type[] targetTypes, object parameter, CultureInfo culture)
+        {
+            throw new NotImplementedException("ConvertBack should NOT be invoked");
```
Nit: You could use `$"{nameof(ConvertBack)} should NOT be invoked"` instead, though it is highly unlikely that the name of this method will ever change. :-)

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/Converters/Converters.cs:
```diff
@@ -0,0 +1,101 @@
+    {
+        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
+        {
+            return !(bool)value;
```
Do we need to guard against the value possibly being something other than a `bool`?

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/Editors/ImplicitUsingsEditor/CSharpImplicitUsingsControl.cs:
```diff
@@ -0,0 +1,394 @@
+        {
+            UsingCollectionState = new ObservableCollection<ImplicitUsingModel>();
+
+#pragma warning disable VSTHRD012
```
It would be good to have a comment explaining why it is OK to disable this warning here.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/Editors/ImplicitUsingsEditor/CSharpImplicitUsingsControl.cs:
```diff
@@ -0,0 +1,394 @@
+            _forceAliasFocus = false;
+            _forceIsStaticFocus = false;
+
+#pragma warning disable VSTHRD012
```
Again, a comment explaining why this is OK here would be good.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/Editors/ImplicitUsingsEditor/CSharpImplicitUsingsControl.cs:
```diff
@@ -0,0 +1,394 @@
+        }
+    }
+
+    internal sealed class ImplicitUsingModel : INotifyPropertyChanged
```
In the "model/view/view-model" pattern is this a model, or a view model?

tmeschter replied:
This is large enough that I would expect to see it in a separate file.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/Editors/PropertyEditorBase.cs:
```diff
@@ -0,0 +1,127 @@
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Implementation.PropertyPages.Designer
+{
+
```
Nit: extra whitespace.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/Editors/PropertyEditorBase.cs:
```diff
@@ -0,0 +1,127 @@
+        {
+            LazyResources = new Lazy<ResourceDictionary>(
+                () =>
+                {
```
Do we need a `UIThreadHelper.VerifyOnUIThread();`?

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ReferencesPage/ImplicitUsingsValueProvider.cs:
```diff
@@ -0,0 +1,200 @@
+namespace Microsoft.VisualStudio.ProjectSystem.Properties;
+
+[ExportInterceptingPropertyValueProvider("ImplicitUsingsEditor", ExportInterceptingPropertyValueProviderFile.ProjectFile)]
+[AppliesTo(ProjectCapability.CSharp)]
```
Note that `AppliesTo` attributes are not currently honored when importing interceptors, so make sure only C# projects use a property by the name of "ImplicitUsingsEditor".

It is still valuable to have to `AppliesTo` attribute in preparation for the day we fix this.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ReferencesPage/ImplicitUsingsValueProvider.cs:
```diff
@@ -0,0 +1,200 @@
+                .GetItems("Using")
+                .Select(item =>
+                {
+                    string? isStaticMetadata = item.DirectMetadata.FirstOrDefault(metadata => metadata.Name.Equals("Static", StringComparison.Ordinal))?.EvaluatedValue;
```
Prefer using one of the values in `StringComparisons` over `StringComparison.Ordinal`/`StringComparison.OrdinalIgnoreCase`/etc.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ReferencesPage/ImplicitUsingsValueProvider.cs:
```diff
@@ -0,0 +1,200 @@
+
+    public override async Task<string?> OnSetPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string>? dimensionalConditions = null)
+    {
+        await _threadingService.SwitchToUIThread();
```
Why do we need the UI thread here?

adamint replied:
Removed. I thought it was necessary.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ReferencesPage/ImplicitUsingsValueProvider.cs:
```diff
@@ -0,0 +1,200 @@
+                catch (Exception ex)
+                {
+                    // if an import comes from a targets file, or else if there's a race condition we can't remove it. otherwise throw
+                    if (ex is not ArgumentException && ex is not InvalidOperationException)
```
I'd prefer an explicit check that the item is not in the project file rather than depending on the lower layers of CPS throwing an exception here.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ReferencesPage/ImplicitUsingsValueProvider.cs:
```diff
@@ -0,0 +1,200 @@
+                hashCode = (hashCode * 397) ^ (Alias != null ? Alias.GetHashCode() : 0);
+                hashCode = (hashCode * 397) ^ IsStatic.GetHashCode();
+                hashCode = (hashCode * 397) ^ IsReadOnly.GetHashCode();
+                return hashCode;
```
We may have a utility type that can help with this.

tmeschter replied:
That is, you may be able to replace all this with a method call.

adamint replied:
It doesn't look like there is :/

tmeschter reviewed (approved):
Generally looks good, modulo some feedback.

I'd also recommend pulling the converters and other WPF-specific utility code into a top level "WPF" folder, just like we have Shell, Telemetry, Threading, etc. They aren't really specific to project system functionality.

MiYanni commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Microsoft.VisualStudio.ProjectSystem.Managed.VS.csproj:
```diff
@@ -88,15 +79,15 @@
 
   <ItemGroup>
     <None Include="**\*.xlf" />
-    <EmbeddedResource Update="ProjectSystem\VS\PropertyPages\PropertyPageResources.resx">
-      <Generator>PublicResXFileCodeGenerator</Generator>
-      <LastGenOutput>PropertyPageResources.Designer.cs</LastGenOutput>
-      <SubType>Designer</SubType>
-    </EmbeddedResource>
```
Here, the .resx uses the generator to, I believe, create the designer files (that were being updated above). We still need this functionality as changes to the .resx would not go into the UI properly (the designer files).

MiYanni commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/Converters/Converters.LambdaConverter.cs:
```diff
@@ -0,0 +1,69 @@
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Utilities
+{
+    // Copied from CPS Converters.cs
+    internal static partial class Converters
```
This seems to be highly generic and not related to the Project System at all. This can either be at the root of the project or an appropriate sub-folder.

adamint replied:
This is only going to be used in the property pages and relates to Converters. I'm of the opinion that it should stay where it is.

MiYanni commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/Converters/Converters.LambdaMultiConverter.cs:
```diff
@@ -0,0 +1,170 @@
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Utilities
+{
+    // Copied from CPS Converters.cs
+    internal static partial class Converters
```
Same comment as above.

adamint replied:
This is only going to be used in the property pages and relates to Converters. I'm of the opinion that it should stay where it is. This is the same thing that CPS does

MiYanni commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/Converters/Converters.cs:
```diff
@@ -0,0 +1,101 @@
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Utilities
+{
+    internal sealed class MultiValueBoolToBool_And : IMultiValueConverter
```
Same comment as above.

adamint replied:
This is only going to be used in the property pages and relates to Converters. I'm of the opinion that it should stay where it is.

MiYanni commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/PropertyPageResources.Designer.cs:
```diff
@@ -22,6 +24,7 @@ namespace Microsoft.VisualStudio.ProjectSystem.VS.PropertyPages {
     [global::System.CodeDom.Compiler.GeneratedCodeAttribute("System.Resources.Tools.StronglyTypedResourceBuilder", "16.0.0.0")]
     [global::System.Diagnostics.DebuggerNonUserCodeAttribute()]
     [global::System.Runtime.CompilerServices.CompilerGeneratedAttribute()]
+    [SuppressMessage("ApiDesign", "RS0041:Public members should not use oblivious types")]
```
What is this for?

adamint commented:
/azp run
